design/20241124-rename-bunde-to-clusterbundle.md

```diff
@@ -0,0 +1,158 @@
+# Design: Renaming Bundle to ClusterBundle
+
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Alternatives](#alternatives)
+- [Future Work](#future-work)
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be
+merged.
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+TODO: add more items to checklist
+
+## Summary
+
+We want to rename `Bundle` to `ClusterBundle` for reasons described in "Motivation" below.
+
+Conversion webhooks in Kubernetes can only be used to convert between versions of and API,
+and not to convert between resources. This makes renaming harder to do, as it will represent
+deleting the old API and creating a new API.
+
+Even though trust-manager `Bundle` API is officially in an alpha state, we want to provide a smooth
+migration path for our users. For this reason we are proposing to do this change over multiple releases,
+requiring end-user action on the way, but without risking trust-manager target resources - which can provide
+critical features in a cluster. For additional details, please see "Proposal" below.
+
+## Motivation
+
+We have [an open issue](https://github.com/cert-manager/trust-manager/issues/63) suggesting renaming Bundle to ClusterBundle.
+Might seem like an obvious WontFix, considering all the assumed work (and pain) to do this,
+but it turns out that the current name somehow blocks introducing an equivalent namespace-scoped resource -
+for trust bundle management inside namespaces in multi-tenant clusters.
+
+In addition, `ClusterBundle` would be a better name looking at cert-manager having `Issuer` and `ClusterIssuer`.
+The name would also be better aligned with upstream [ClusterTrustBundle](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#cluster-trust-bundles).
+
+### Goals
+
+- `Bundle` resource is renamed to `ClusterBundle`
+- Provide a smooth migration for users of trust-manager
+
+
+### Non-Goals
+
+- Change/fix/improve `Bundle`/`ClusterBundle` API while renaming
```

Comment:
If there are already known improvements, there's no particular reason to block them during this rename -- you can add them to the new ClusterBundle resource and adapt the Bundle shape as part of the migration.

Reply:
True! One example might be to make the namespaceSelector required. But any change will add complexity to the migration controller.

Reply:
I think the complexity will be similar to a storage conversion controller, but it does add some complexity.

Reply:
Thanks a lot for pushing me on this @evankanderson! 👍 It seems to actually represent a huge opportunity for trust-manager. I did a simple draft of proposed API changes in this PR: https://github.com/cert-manager/trust-manager/pull/486. Will use it when discussing the proposal in our next bi-weekly meeting.

Reply:
Ok, I believe we can gain something if improving the API when migrating. I've now made adjustments to the proposal.